### DISCUSSION TITLE
TailwindCSS + Next.js

### DISCUSSION THREAD

**Discussion by blade2201** [vercel/next.js#38593] [Help] [answered]
I just started getting into Next.js and was wondering if using tailwind you loose some of the pros of writing css modules.

Meaning for example that with tailwind you only get a compiled css file from with all the purged classes in your project that will be loaded on every route while with css modules you could create route specific css that will only get loaded only when required.

Is this correct? Should i care about loosing this efficiency for a small solo project?

**Comment by icyJoseph** [comment]
You know, instinctively I wanna say that it should NOT be a problem, at least not unless all pages are, strictly, unique in terms of style, because, most of the Tailwind generated CSS is highly shareable between pages. 

For example I released a site last week and the total CSS as measured by https://cssstats.com/ shows 5kB, in all pages of course as you note, but the thing is that using this other tool https://developer.chrome.com/docs/devtools/coverage/, I see that the worst case scenario is a page with 54% unused CSS (of 5kB), most of which is CSS resets/normalise injected by Tailwind, because I am not using every possible available HTML tag. Perhaps I can try to optimise those out. 

The page that uses most of the CSS is at about 76%. 

So, yeah, I think it shouldn't be a big deal, but it is good to consider for sure, because not only will there be a single CSS file, as far as I can see, and understand, but it won't be used to the fullest extend by all pages.

I guess a smaller project, where each page is very unique, might be affected by this more than a large app, that shares more and more components between each page. Of course, speaking of percentages of unused CSS per page. Then again, once this file is in the browser, all other pages benefit from it, AKA, no more downloads of CSS are necessary.

For a small solo project, I think it shouldn't be a problem, but rather an opportunity to learn and study how Tailwind CSS and Next.js, scale and what techniques one could use to cut down in unused CSS.

**Reply by blade2201** [reply]
Wow, thanks for the nice explanation.

**Comment by GabenGar** [comment, marked as answer]
Tailwind is a massive meme. For its purging ability to work and not break you have to basically write all classnames as string literals without interpolation, otherwise it will get confused on what to purge or not.
Also its theme implementation is terrible, you have to basically double the amount of declarations everywhere to support light and dark theme (remember you can't interpolate the classname). I've ended up in a situation of writing 10+ rules to style a link element. And for child selectors you have to install a separate package without a major version.
Nextjs-specific issue especially for smaller projects is it introduces 2+ configs, one of them is a postcss config, so it's 2 more configs to manage right off the bat.

**Reply by icyJoseph** [reply]
Might be good to also add, that when you go TW, you deny IE11 support. CSS custom properties and grid are a big no go. 

And sure IE11 should be abandoned and whatever but if the client requires IE support, then that's it.

**Reply by GabenGar** [reply]
It's not really a problem unless absolutely required for the project. But also is a Tailwind issue as it does use CSS custom properties extensively, but you can't use them meaningfully while using it.
